services/director-v2/src/simcore_service_director_v2/utils/dags.py

```diff
@@ -198,12 +198,15 @@ async def compute_pipeline_details(
 
 
 @log_decorator(logger=logger)
-def topological_sort_grouping(dag_graph: nx.DiGraph) -> List:
+def topological_sort_grouping(dag_graph: nx.DiGraph) -> List[Dict[str, Dict[str, Any]]]:
```

Comment:
Do i understand this correctly, all nodes with zero incoming edges are added to the same group and as a result are executed serial? Or do you create one group for all of them so that they can run in parallel

Reply:
right.
1. All nodes with zero incoming edges go in one List (group0), then are popped out, then the next List (group1), etc...
2. After that I parse these and put them in celery "chords" which are run in parallel. A chord is a group of tasks *run in parallel* that then trigger a following task which is itself another chord.
```chord(header=group0, body=chord(header=group1, body=chord(header=group2, body=...groupN)))```

Reply:
for reference [Chords](https://docs.celeryproject.org/en/stable/userguide/canvas.html#chords)

Reply:
So all nodes in group0 run serial?

Reply:
Ah. I see. They run in parallel. Very good.

Reply:
MINOR: Should you add this explanation in the code? it is not clear from the outside what is the purpose of grouping

SMALL: should you add a test for this function in  ``services/director-v2/tests/unit/test_dags.py`` which BTW should be instead called ``services/director-v2/tests/unit/test_utils_dags.py``.

Reply:
test added. and documentation modified